docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,211 @@
+- Defining the changes needed in LoRa Device Provisioning CLI for allowing
+  firmware upgrades
+
+## Out-of-scope
+
+- Generating of the signature key, CRC32 checksum of the signature and digest in
+  the LoRa Device Provisioning CLI is considered a stretch and will not be added
+  to the tool as a functionality for the time being. This document provides
+  sample commands which can be used to generate required values in the appendix.
+
+## Decisions
+
+### Context
+
+The CUPS request described in [The CUPS protocol - documentation][cupsproto]
+contains the `package` field which indicates the current firmware version of the
+Basics Station. The value will need to be updated whenever a firmware upgrade of
+the Basics Station is performed by the user.
+
+### Firmware upgrade flow
+
+1. When provisioning a device that will need firmware upgrades, the end user
+   should generate a `sig-0.key` and store it on the device and in a centralized
+   repository of their own choice (example on how to generate it can be found in
+   the appendix).
+
+1. When a firmware upgrade needs to be provided to the device, the user should
+   generate a digest of the executable file of the upgrade (as in examples
+   mentioned above) and retrieve the CRC32 Checksum of the `sig-0.key`. Then 4
+   inputs (file, digest, checksum and new version) should be provided to the
+   LoRa Device Provisioning CLI tool to properly upload the file to a blob
+   storage and store the needed information in the concentrator twin (blob URL
+   in storage, digest, checksum and the new version number).
+
+1. The Basics Station sends the CUPS request containing the currently used
+   version (in the `package` field).
+
+1. The LNS compares the values from the device twin and the CUPS request, and
+   determines if an upgrade is required. If that's the case, the LNS will
+   download the upgrade file from storage and send a properly populated response
+   to the Basics Station.
+
+1. The Basics Station will then execute the actual firmware upgrade.
+
+```mermaid
+  sequenceDiagram
+    autonumber
+    User->>LoRa Device Provisioning CLI: Request Station upgrade
+    LoRa Device Provisioning CLI->>Storage account: Upload firmware file
+    LoRa Device Provisioning CLI->>IoT Hub: Update 'Concentrator' twin
+    Concentrator->>Network Server: POST /update-info
+    Network Server->>IoT Hub: Retrieve 'Concentrator' twin
+    Network Server->>Network Server: Check if versions of the Station are different
+    alt versions are the same
+        Network Server->>Concentrator: CUPS response without firmware upgrade
+    else versions are different
+      Network Server->>Network Server: Verify if any of the `keys` from CUPS request match the `fwKeyChecksum`
+      alt no match
+        Network Server->>Network Server: Throw as this indicates a misconfiguration
+      else one of the keys matches `fwKeyChecksum`
+        Network Server->>Facade Function: Retrieve firmware upgrade file
```

Comment:
How big do we expect that these firmware upgrade files can be? Will there be potential restrictions to the size of the firmware upgrade package through the timeout of the Azure Function at lower network speeds?